File: backend/src/main/java/com/roadmap/Roadmap/controllers/CourseData.java
package com.roadmap.Roadmap.controllers;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.roadmap.Roadmap.models.Course;
import com.roadmap.Roadmap.models.Course.Topic;
import com.roadmap.Roadmap.models.Course.Topic.Subtopic;
import com.roadmap.Roadmap.repositories.CourseRepo;

@RestController
@RequestMapping("/api/private")
@CrossOrigin(origins = "*")
public class CourseData {
    @Autowired
    CourseRepo Courses;

    @GetMapping("/getCourse")
    @CrossOrigin(origins = "*")
    public Course getCourseById(@RequestParam String id) {
        return Courses.findById(id).orElse(null);
    }

    @PostMapping("/addNewCourse")
    @CrossOrigin(origins = "*")
    public Course addNewCourse(@RequestBody Course course) {
        return Courses.save(course);
    }

    @PatchMapping("/updateCourse")
    @CrossOrigin(origins = "*")
    public Course updateCourse(@RequestBody Course course) {
        return Courses.save(course);
    }

    @PatchMapping("/addNewTopic") // can also be used for updating the topic
    @CrossOrigin(origins = "*")
    public Course addNewTopic(@RequestParam String id, @RequestBody Topic topic) {
        Course course = Courses.findById(id).orElse(null);
        course.topics.put(topic.name, topic);
        return course;
    }

    @PatchMapping("/deleteTopic")
    @CrossOrigin(origins = "*")
    public Course deleteTopic(@RequestParam String id, @RequestBody String topicName) {
        Course course = Courses.findById(id).orElse(null);
        course.topics.remove(topicName);
        return course;
    }

    public class AddNewSubtopicReqBody {
        String topicName;
        Subtopic subtopic;
    }

    @PatchMapping("/addNewSubtopic")
    @CrossOrigin(origins = "*")
    public Course addNewSubtopic(@RequestParam String id, @RequestBody AddNewSubtopicReqBody body) {
        Course course = Courses.findById(id).orElse(null);
        course.topics.get(body.topicName).subtopics.put(body.subtopic.name, body.subtopic);
        return course;
    }

    public class DeleteSubtopicReqBody {
        String topicName;
        String subtopicName;
    }

    @PatchMapping("/deleteSubtopic")
    @CrossOrigin(origins = "*")
    public Course deleteSubtopic(@RequestParam String id, @RequestBody DeleteSubtopicReqBody body) {
        Course course = Courses.findById(id).orElse(null);
        course.topics.get(body.topicName).subtopics.remove(body.subtopicName);
        return course;
    }
}

File: backend/controllers/CourseData.js
const express = require("express");
const router = express.Router();
const Course = require("../models/Course");

// Import the CourseRepo if you're using one for database operations

// GET course by ID
router.get("/getCourse", async (req, res) => {
  try {
    const course = await Course.findById(req.query.id);
    res.json(course);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// POST add new course
router.post("/addNewCourse", async (req, res) => {
  try {
    const course = new Course({
      name: req.body.name,
      topics: req.body.topics,
    });
    const newCourse = await course.save();
    res.status(201).json(newCourse);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// PATCH update course
router.patch("/updateCourse", async (req, res) => {
  try {
    const updatedCourse = await Course.findByIdAndUpdate(
      req.body.id,
      req.body,
      { new: true }
    );
    res.json(updatedCourse);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// PATCH add new topic or update existing topic
router.patch("/addNewTopic", async (req, res) => {
  try {
    const course = await Course.findById(req.query.id);
    course.topics.set(req.body.name, req.body);
    const updatedCourse = await course.save();
    res.json(updatedCourse);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// PATCH delete topic
router.patch("/deleteTopic", async (req, res) => {
  try {
    const course = await Course.findById(req.query.id);
    course.topics.delete(req.body);
    const updatedCourse = await course.save();
    res.json(updatedCourse);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// PATCH add new subtopic
router.patch("/addNewSubtopic", async (req, res) => {
  try {
    const course = await Course.findById(req.query.id);
    course.topics
      .get(req.body.topicName)
      .subtopics.set(req.body.subtopic.name, req.body.subtopic);
    const updatedCourse = await course.save();
    res.json(updatedCourse);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// PATCH delete subtopic
router.patch("/deleteSubtopic", async (req, res) => {
  try {
    const course = await Course.findById(req.query.id);
    course.topics
      .get(req.body.topicName)
      .subtopics.delete(req.body.subtopicName);
    const updatedCourse = await course.save();
    res.json(updatedCourse);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

module.exports = router;
